projecttacoma/bulk-export-app
# Implement Date Type Filter Selection

# Summary

Create ability to add type filters with a dropdown menu and a modal selection window.

## Issue

There is an error in the console, when you reopen the type filter modal by clicking the icon in the table, and then try to select a date again. It is an accessibility error that I have been trying to fix, but it doesn't effect the functionality of the modal. It seems to be caused by the opening of the date time selector modal inside the other modal.

## New behavior

On the query builder page, there is now an dropdown to add a type filter. You can select a type to add a filter on. This opens a modal that is similar to the elements modal where you can select a number of search parameters to add filters on. You can add some search parameters which will create input boxes for each of them below. By default, the input is a text input, but for the search parameters that are date types, the input field is a Mantine DateTimePicker as well as a comparator select input. After you enter the values for the search parameters that you want to include on the type filer, you can press the confirm button, which will add the type filter to the table.

You can edit type filters by pressing the edit icon in the table, which will reopen the modal with the fields already populated with information selected previously.

## Code changes
- `package.json` new fhir-spec-version and added the Mantine date package.
- `src/app/providers.tsx` added the Mantine date provider, along with its styles, added styles for Mantine select components.
- `src/components/parameter-pages/type-filter-modal.tsx` new component for the modal content to create type filters
- `src/components/parameter-pages/type-filter-param-input.tsx` component to input values for each search parameter.
- `src/components/parameter-pages/type-filter-params-page.tsx` added selection to select a type that will open the modal to create a type filter
- `src/components/parameter-pages/type-filter-table.tsx` added button to edit a type filter which opens the modal as well as a function to create TypeFilterParamInputs for each of the search parameters in the type filter.

# Testing guidance

- `npm run check`
- `npm run build`
- `npm run start`
- Navigate through the pages as usual and go to the type filter tab on the query builder page
- Click the selection box below the text input for type filters
- Search and select a type to add a filter to, which opens a modal
- Add search parameters in the Multiselect at the top of the modal window
- Add values to the input fields for each parameter, or use the date time picker for date type search parameters
- Click the confirm button and verify that the filter appears correctly in the table
- Click the edit button for that filter, and verify that it reopens the modal window with the input fields set.
- When you have selected some filters, navigate to the next page and check that the filters applied to the export request correctly. `patient?gender=female` is a good test for this
- As for testing the date parameters, because of the issue with the bulk export server, this is a little challenging, but if you look for a specific resource's creationDate or some other form of date, you can select a date that is close to that and test the 'ap' comparator, or select the exact date and test the 'eq' comparator. I believe these are the only two that are working as intended at the moment.
